presto-main/src/main/java/com/facebook/presto/sql/planner/TranslationMap.java

```diff
@@ -131,14 +136,23 @@ public Symbol get(Expression expression)
     {
         if (expression instanceof FieldReference) {
             int field = ((FieldReference) expression).getFieldIndex();
-            Preconditions.checkArgument(fieldSymbols[field] != null, "No mapping for field: %s", field);
+            checkArgument(fieldSymbols[field] != null, "No mapping for field: %s", field);
             return fieldSymbols[field];
         }
 
         Expression translated = translateNamesToSymbols(expression);
+        checkArgument(expressionToSymbols.containsKey(translated), "No mapping for expression: %s", expression);
+        return expressionToSymbols.get(translated);
+    }
 
-        Preconditions.checkArgument(expressionMappings.containsKey(translated), "No mapping for expression: %s", expression);
-        return expressionMappings.get(translated);
+    public void put(Expression expression, Expression rewritten)
+    {
+        expressionToExpressions.put(translateNamesToSymbols(expression), rewritten);
+    }
+
+    public void setExpressionAsAlreadyTranslated(Expression node)
```

Comment:
Using `IdentityHashMap` didn't work out. `TranslationMap` first translate names to symbols. This always creates a new instance, so `IdentityHashMap` does not work in such case.